set-static-ip

```diff
@@ -1,20 +1,26 @@
 #!/bin/bash -xe
 
 #PROVISIONING_INTERFACE='ens3'
-#PROVISIONING_IP='172.22.0.2'
-
-if [ -z "$PROVISIONING_INTERFACE" ]; then
-    echo "ERROR: PROVISIONING_INTERFACE environment variable unset."
-    exit 1
-fi
+#PROVISIONING_IP='172.22.0.2/24'
 if [ -z "$PROVISIONING_IP" ]; then
     echo "ERROR: PROVISIONING_IP environment variable unset."
     exit 1
 fi
 
-# Get rid of any DHCP addresses etc.
-/usr/sbin/ip address flush dev $PROVISIONING_INTERFACE
+if [ -z "$PROVISIONING_INTERFACE" ]; then
+  # If no provisioning interface is specified, then we're probably in
+  # the mode where the provisioning network is optional, so to avoid a
+  # hard dependency of forcing users to specify the interface, we detect
+  # which network interface has the IP's from the machine network.
+  IP_ONLY=$(echo "$PROVISIONING_IP" | cut -d/ -f1)
+  PROVISIONING_INTERFACE=$(ip route get "$IP_ONLY" | grep -Po '(?<=(dev )).*(?= src| proto)')
```

Comment:
If the pod gets restarted and the IP is already configured on an interface, this will return `lo` I think, so I wonder if we should flush before this step to avoid getting the wrong interface?

Reply:
Flushing would get rid of the DHCP lease though, and probably break the kubelet :-\

Reply:
I'll have a look once I get to a point where I can test some stuff together

Reply:
Yeah I realized after posting that flushing was a bad idea :)  I guess we could detect the case where the IP is already up and either take it down, or use the interface name that it's configured on?

Reply:
WDYT about checking if `$PROVISIONING_NETWORK` is `disabled` and letting the user to specify `$PROVISIONING_INTERFACE` in any case?

It feels less error-prone to me.

Reply:
Also, there are other consumers for `PROVISIONING_INTERFACE` IIRC (BMO for example). Are they all can work with an empty value?

Reply:
> WDYT about checking if $PROVISIONING_NETWORK is disabled and letting the user to specify $PROVISIONING_INTERFACE in any case?

What's the use case? I don't want to have a million different ways to configure baremetal, I don't see a reason for it. We *know* the machine CIDR the VIP will be on, so we can find the interface.

> Also, there are other consumers for PROVISIONING_INTERFACE IIRC (BMO for example). Are they all can work with an empty value?

They all fallback to the IP.

Reply:
@stbenjam 
> > WDYT about checking if $PROVISIONING_NETWORK is disabled and letting the user to specify $PROVISIONING_INTERFACE in any case?
> 
> What's the use case?

use case - a cluster with a single NIC on each server (I was told that 2 NICs are not mandatory), where `ProvisioningNetwork` is set to `Disabled` in `Provisioning` CR

>  I don't want to have a million different ways to configure baremetal, I don't see a reason for it. We _know_ the machine CIDR the VIP will be on, so we can find the interface.

I wasn't suggesting to add a new config flag, rather using the new provisioning_network field, as described in 
[this proposal](https://github.com/openshift/enhancements/blob/master/enhancements/baremetal/baremetal-provisioning-optional.md#proposal)

I just think that flushing interfaces can be risky, flushing `lo` interface will result in many failure, so I thought that working in explicit manner could be better. but you guys are more familiar with this area, so whatever works for you :+1: